include/lm/language_model.h

```diff
@@ -26,87 +51,74 @@ class language_model
+    float perplexity(const sentence& tokens) const;
 
     /**
-     * @param tokens A sequence of tokens
+     * @param sentence A sequence of tokens
      * @return the perplexity of this token sequence given the current language
      * model normalized by the length of the sequence
      */
-    double perplexity_per_word(const std::string& tokens) const;
+    float perplexity_per_word(const sentence& tokens) const;
 
     /**
-     * @param tokens A sequence of n tokens
-     * @return the probability of seeing the nth token based on the previous n
-     * - 1 tokens
+     * @param tokens A sequence of n tokens (one sentence)
+     * @return the log probability of the likelihood of this sentence
      */
-    double prob(std::deque<std::string> tokens) const;
+    float log_prob(sentence tokens) const;
 
-  private:
+    /**
+     * @param prev Seen tokens to base the next token off of
+     * @param k Number of results to return
+     * @return a sorted vector of likely next tokens
+     */
+    std::vector<std::pair<std::string, float>> top_k(const sentence& prev,
+                                                     size_t k) const;
 
+  private:
     /**
-     * Builds the probabilities associated with this language model.
-     * @param config_file The config file that specifies the location of the
-     * corpus
+     * Reads precomputed LM data into this object.
+     * @param arpa_file The path to the ARPA-formatted file
      */
-    void learn_model(const std::string& config_file);
+    void read_arpa_format(const std::string& arpa_file);
 
     /**
-     * @param tokens A deque of tokens to convert to a string
-     * @return the string version of the deque (space delimited)
+     * @param tokens
+     * @return the log probability of one ngram
      */
-    std::string make_string(const std::deque<std::string>& tokens) const;
+    float prob_calc(sentence tokens) const;
 
     /**
-     * @param tokens A string of space-delimited tokens to convert to a deque
-     * @return a deque of the tokens
+     * Loads unigram vocabulary from text file to allow top_k to work.
      */
-    std::deque<std::string> make_deque(const std::string& tokens) const;
+    void load_vocab();
 
-    /// The language_model used to interpolate with this one for smoothing
-    std::unique_ptr<language_model> interp_;
+    uint64_t N_; /// The "n" value for this n-gram language model
 
-    /// Contains the N-gram distribution probabilities (N-1 words -> (w, prob))
-    std::unordered_map
-        <std::string, std::unordered_map<std::string, double>> dist_;
+    std::vector<static_probe_map> lm_;
 
-    /// The value of N in this n-gram
-    size_t N_;
+    std::vector<std::string> vocabulary_;
 
-    /// The interpolation coefficient for smoothing LM probabilities
-    constexpr static double lambda_ = 0.7;
+    std::string prefix_;
+};
+
+class language_model_exception : public std::runtime_error
+{
+    using std::runtime_error::runtime_error;
 };
```

Comment:
I don't know what our stance is on this in general, but I prefer to have things like `language_model::exception` if possible (but that gets cumbersome with template classes).

Whatever we decide, we should probably make this consistent across the toolkit, since I don't think it currently is.